File: artifacts/avl-tree/others/recode/tool_projects/orbit/avl-tree/c/avl_data_test_generated.c
/*
 * Generated test file for print_char_func function
 * Tests the character printing functionality of avl_data
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include "avl_data.h"

// Helper function to capture stdout output
char captured_output[1024];
int capture_pos = 0;

void reset_capture() {
    memset(captured_output, 0, sizeof(captured_output));
    capture_pos = 0;
}

void capture_char(char c) {
    if (capture_pos < sizeof(captured_output) - 1) {
        captured_output[capture_pos++] = c;
        captured_output[capture_pos] = '\0';
    }
}

// Custom print_char_func that captures output for testing
void test_print_char_func(void *d) {
    mydata *p;

    assert(d != NULL);

    p = (mydata *) d;
    char c = (char)(p->key & 127);
    capture_char(c);
}

// Test case structure
typedef struct {
    int key;
    char expected_char;
    const char *description;
} test_case_t;

// Test basic printable ASCII characters
void test_uppercase_letters() {
    printf("Test: Uppercase letters... ");

    test_case_t cases[] = {
        {65, 'A', "Letter A"},
        {90, 'Z', "Letter Z"},
        {77, 'M', "Letter M"}
    };

    for (int i = 0; i < 3; i++) {
        reset_capture();
        mydata *data = makedata(cases[i].key);
        test_print_char_func(data);

        if (captured_output[0] != cases[i].expected_char) {
            printf("FAILED: %s - Expected '%c', got '%c'\n",
                   cases[i].description, cases[i].expected_char, captured_output[0]);
            destroy_func(data);
            exit(1);
        }
        destroy_func(data);
    }

    printf("PASSED\n");
}

// Test lowercase letters
void test_lowercase_letters() {
    printf("Test: Lowercase letters... ");

    test_case_t cases[] = {
        {97, 'a', "Letter a"},
        {122, 'z', "Letter z"},
        {109, 'm', "Letter m"}
    };

    for (int i = 0; i < 3; i++) {
        reset_capture();
        mydata *data = makedata(cases[i].key);
        test_print_char_func(data);

        if (captured_output[0] != cases[i].expected_char) {
            printf("FAILED: %s - Expected '%c', got '%c'\n",
                   cases[i].description, cases[i].expected_char, captured_output[0]);
            destroy_func(data);
            exit(1);
        }
        destroy_func(data);
    }

    printf("PASSED\n");
}

// Test digits
void test_digits() {
    printf("Test: Digit characters... ");

    test_case_t cases[] = {
        {48, '0', "Digit 0"},
        {57, '9', "Digit 9"},
        {53, '5', "Digit 5"}
    };

    for (int i = 0; i < 3; i++) {
        reset_capture();
        mydata *data = makedata(cases[i].key);
        test_print_char_func(data);

        if (captured_output[0] != cases[i].expected_char) {
            printf("FAILED: %s - Expected '%c', got '%c'\n",
                   cases[i].description, cases[i].expected_char, captured_output[0]);
            destroy_func(data);
            exit(1);
        }
        destroy_func(data);
    }

    printf("PASSED\n");
}

// Test special characters
void test_special_characters() {
    printf("Test: Special characters... ");

    test_case_t cases[] = {
        {32, ' ', "Space"},
        {33, '!', "Exclamation"},
        {64, '@', "At symbol"},
        {126, '~', "Tilde"}
    };

    for (int i = 0; i < 4; i++) {
        reset_capture();
        mydata *data = makedata(cases[i].key);
        test_print_char_func(data);

        if (captured_output[0] != cases[i].expected_char) {
            printf("FAILED: %s - Expected '%c', got '%c'\n",
                   cases[i].description, cases[i].expected_char, captured_output[0]);
            destroy_func(data);
            exit(1);
        }
        destroy_func(data);
    }

    printf("PASSED\n");
}

// Test masking behavior (key & 127)
void test_masking_behavior() {
    printf("Test: Masking behavior (key & 127)... ");

    test_case_t cases[] = {
        {128, '\0', "128 masked to 0"},
        {129, '\x01', "129 masked to 1"},
        {193, 'A', "193 masked to 65 (A)"},
        {255, '\x7f', "255 masked to 127"}
    };

    for (int i = 0; i < 4; i++) {
        reset_capture();
        mydata *data = makedata(cases[i].key);
        test_print_char_func(data);

        if (captured_output[0] != cases[i].expected_char) {
            printf("FAILED: %s - Expected 0x%02x, got 0x%02x\n",
                   cases[i].description, (unsigned char)cases[i].expected_char,
                   (unsigned char)captured_output[0]);
            destroy_func(data);
            exit(1);
        }
        destroy_func(data);
    }

    printf("PASSED\n");
}

// Test edge cases
void test_edge_cases() {
    printf("Test: Edge cases... ");

    test_case_t cases[] = {
        {0, '\0', "Null character"},
        {127, '\x7f', "DEL character (127)"},
        {1, '\x01', "SOH character"}
    };

    for (int i = 0; i < 3; i++) {
        reset_capture();
        mydata *data = makedata(cases[i].key);
        test_print_char_func(data);

        if (captured_output[0] != cases[i].expected_char) {
            printf("FAILED: %s - Expected 0x%02x, got 0x%02x\n",
                   cases[i].description, (unsigned char)cases[i].expected_char,
                   (unsigned char)captured_output[0]);
            destroy_func(data);
            exit(1);
        }
        destroy_func(data);
    }

    printf("PASSED\n");
}

int main() {
    printf("=== Testing print_char_func ===\n\n");

    test_uppercase_letters();
    test_lowercase_letters();
    test_digits();
    test_special_characters();
    test_masking_behavior();
    test_edge_cases();

    printf("\n=== All tests PASSED ===\n");
    return 0;
}

File: artifacts/avl-tree/others/recode/tool_projects/orbit/avl-tree/rust/src/bin/avl_data_test_generated.rs
/*
 * Generated test file for print_char_func function
 * Tests the character printing functionality of avl_data
 */

use std::process;

// Import the avl_data module
// Assuming the crate name is 'avl_tree' based on typical Rust project structure
extern crate avl_tree;
use avl_tree::avl_data::{makedata, MyData};

// Helper struct to capture output
static mut CAPTURED_OUTPUT: String = String::new();

fn reset_capture() {
    unsafe {
        CAPTURED_OUTPUT.clear();
    }
}

fn capture_char(c: char) {
    unsafe {
        CAPTURED_OUTPUT.push(c);
    }
}

// Custom print_char_func that captures output for testing
fn test_print_char_func(d: &MyData) {
    let c = (d.key & 127) as u8 as char;
    capture_char(c);
}

// Test case structure
struct TestCase {
    key: i32,
    expected_char: char,
    description: &'static str,
}

// Test basic printable ASCII characters
fn test_uppercase_letters() {
    print!("Test: Uppercase letters... ");

    let cases = vec![
        TestCase { key: 65, expected_char: 'A', description: "Letter A" },
        TestCase { key: 90, expected_char: 'Z', description: "Letter Z" },
        TestCase { key: 77, expected_char: 'M', description: "Letter M" },
    ];

    for case in cases {
        reset_capture();
        let data = makedata(case.key);
        test_print_char_func(&data);

        unsafe {
            if CAPTURED_OUTPUT.chars().next() != Some(case.expected_char) {
                println!(
                    "FAILED: {} - Expected '{}', got '{}'",
                    case.description,
                    case.expected_char,
                    CAPTURED_OUTPUT.chars().next().unwrap_or('\0')
                );
                process::exit(1);
            }
        }
    }

    println!("PASSED");
}

// Test lowercase letters
fn test_lowercase_letters() {
    print!("Test: Lowercase letters... ");

    let cases = vec![
        TestCase { key: 97, expected_char: 'a', description: "Letter a" },
        TestCase { key: 122, expected_char: 'z', description: "Letter z" },
        TestCase { key: 109, expected_char: 'm', description: "Letter m" },
    ];

    for case in cases {
        reset_capture();
        let data = makedata(case.key);
        test_print_char_func(&data);

        unsafe {
            if CAPTURED_OUTPUT.chars().next() != Some(case.expected_char) {
                println!(
                    "FAILED: {} - Expected '{}', got '{}'",
                    case.description,
                    case.expected_char,
                    CAPTURED_OUTPUT.chars().next().unwrap_or('\0')
                );
                process::exit(1);
            }
        }
    }

    println!("PASSED");
}

// Test digits
fn test_digits() {
    print!("Test: Digit characters... ");

    let cases = vec![
        TestCase { key: 48, expected_char: '0', description: "Digit 0" },
        TestCase { key: 57, expected_char: '9', description: "Digit 9" },
        TestCase { key: 53, expected_char: '5', description: "Digit 5" },
    ];

    for case in cases {
        reset_capture();
        let data = makedata(case.key);
        test_print_char_func(&data);

        unsafe {
            if CAPTURED_OUTPUT.chars().next() != Some(case.expected_char) {
                println!(
                    "FAILED: {} - Expected '{}', got '{}'",
                    case.description,
                    case.expected_char,
                    CAPTURED_OUTPUT.chars().next().unwrap_or('\0')
                );
                process::exit(1);
            }
        }
    }

    println!("PASSED");
}

// Test special characters
fn test_special_characters() {
    print!("Test: Special characters... ");

    let cases = vec![
        TestCase { key: 32, expected_char: ' ', description: "Space" },
        TestCase { key: 33, expected_char: '!', description: "Exclamation" },
        TestCase { key: 64, expected_char: '@', description: "At symbol" },
        TestCase { key: 126, expected_char: '~', description: "Tilde" },
    ];

    for case in cases {
        reset_capture();
        let data = makedata(case.key);
        test_print_char_func(&data);

        unsafe {
            if CAPTURED_OUTPUT.chars().next() != Some(case.expected_char) {
                println!(
                    "FAILED: {} - Expected '{}', got '{}'",
                    case.description,
                    case.expected_char,
                    CAPTURED_OUTPUT.chars().next().unwrap_or('\0')
                );
                process::exit(1);
            }
        }
    }

    println!("PASSED");
}

// Test masking behavior (key & 127)
fn test_masking_behavior() {
    print!("Test: Masking behavior (key & 127)... ");

    let cases = vec![
        TestCase { key: 128, expected_char: '\0', description: "128 masked to 0" },
        TestCase { key: 129, expected_char: '\x01', description: "129 masked to 1" },
        TestCase { key: 193, expected_char: 'A', description: "193 masked to 65 (A)" },
        TestCase { key: 255, expected_char: '\x7f', description: "255 masked to 127" },
    ];

    for case in cases {
        reset_capture();
        let data = makedata(case.key);
        test_print_char_func(&data);

        unsafe {
            if CAPTURED_OUTPUT.chars().next() != Some(case.expected_char) {
                println!(
                    "FAILED: {} - Expected 0x{:02x}, got 0x{:02x}",
                    case.description,
                    case.expected_char as u8,
                    CAPTURED_OUTPUT.chars().next().unwrap_or('\0') as u8
                );
                process::exit(1);
            }
        }
    }

    println!("PASSED");
}

// Test edge cases
fn test_edge_cases() {
    print!("Test: Edge cases... ");

    let cases = vec![
        TestCase { key: 0, expected_char: '\0', description: "Null character" },
        TestCase { key: 127, expected_char: '\x7f', description: "DEL character (127)" },
        TestCase { key: 1, expected_char: '\x01', description: "SOH character" },
    ];

    for case in cases {
        reset_capture();
        let data = makedata(case.key);
        test_print_char_func(&data);

        unsafe {
            if CAPTURED_OUTPUT.chars().next() != Some(case.expected_char) {
                println!(
                    "FAILED: {} - Expected 0x{:02x}, got 0x{:02x}",
                    case.description,
                    case.expected_char as u8,
                    CAPTURED_OUTPUT.chars().next().unwrap_or('\0') as u8
                );
                process::exit(1);
            }
        }
    }

    println!("PASSED");
}

fn main() {
    println!("=== Testing print_char_func ===\n");

    test_uppercase_letters();
    test_lowercase_letters();
    test_digits();
    test_special_characters();
    test_masking_behavior();
    test_edge_cases();

    println!("\n=== All tests PASSED ===");
}
